// eslint-plugin-testing-library/rules/no-promise-in-fire-event.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RULE_NAME = void 0;
const experimental_utils_1 = require("@typescript-eslint/experimental-utils");
const create_testing_library_rule_1 = require("../create-testing-library-rule");
const node_utils_1 = require("../node-utils");
exports.RULE_NAME = 'no-promise-in-fire-event';
exports.default = create_testing_library_rule_1.createTestingLibraryRule({
    name: exports.RULE_NAME,
    meta: {
        type: 'problem',
        docs: {
            description: 'Disallow the use of promises passed to a `fireEvent` method',
            category: 'Best Practices',
            recommendedConfig: {
                dom: 'error',
                angular: 'error',
                react: 'error',
                vue: 'error',
            },
        },
        messages: {
            noPromiseInFireEvent: "A promise shouldn't be passed to a `fireEvent` method, instead pass the DOM element",
        },
        schema: [],
    },
    defaultOptions: [],
    create(context, _, helpers) {
        function checkSuspiciousNode(node, originalNode) {
            if (experimental_utils_1.ASTUtils.isAwaitExpression(node)) {
                return;
            }
            if (node_utils_1.isNewExpression(node)) {
                if (node_utils_1.isPromiseIdentifier(node.callee)) {
                    context.report({
                        node: originalNode !== null && originalNode !== void 0 ? originalNode : node,
                        messageId: 'noPromiseInFireEvent',
                    });
                    return;
                }
            }
            if (node_utils_1.isCallExpression(node)) {
                const domElementIdentifier = node_utils_1.getDeepestIdentifierNode(node);
                if (!domElementIdentifier) {
                    return;
                }
                if (helpers.isAsyncQuery(domElementIdentifier) ||
                    node_utils_1.isPromiseIdentifier(domElementIdentifier)) {
                    context.report({
                        node: originalNode !== null && originalNode !== void 0 ? originalNode : node,
                        messageId: 'noPromiseInFireEvent',
                    });
                    return;
                }
            }
            if (experimental_utils_1.ASTUtils.isIdentifier(node)) {
                const nodeVariable = experimental_utils_1.ASTUtils.findVariable(context.getScope(), node.name);
                if (!nodeVariable) {
                    return;
                }
                for (const definition of nodeVariable.defs) {
                    const variableDeclarator = definition.node;
                    if (variableDeclarator.init) {
                        checkSuspiciousNode(variableDeclarator.init, node);
                    }
                }
            }
        }
        return {
            'CallExpression Identifier'(node) {
                if (!helpers.isFireEventMethod(node)) {
                    return;
                }
                const closestCallExpression = node_utils_1.findClosestCallExpressionNode(node, true);
                if (!closestCallExpression) {
                    return;
                }
                const domElementArgument = closestCallExpression.arguments[0];
                checkSuspiciousNode(domElementArgument);
            },
        };
    },
});
