// eslint-plugin-testing-library/rules/prefer-explicit-assert.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RULE_NAME = void 0;
const experimental_utils_1 = require("@typescript-eslint/experimental-utils");
const create_testing_library_rule_1 = require("../create-testing-library-rule");
const node_utils_1 = require("../node-utils");
const utils_1 = require("../utils");
exports.RULE_NAME = 'prefer-explicit-assert';
const isAtTopLevel = (node) => { var _a; return !!((_a = node.parent) === null || _a === void 0 ? void 0 : _a.parent) && node.parent.parent.type === 'ExpressionStatement'; };
exports.default = create_testing_library_rule_1.createTestingLibraryRule({
    name: exports.RULE_NAME,
    meta: {
        type: 'suggestion',
        docs: {
            description: 'Suggest using explicit assertions rather than just `getBy*` queries',
            category: 'Best Practices',
            recommendedConfig: {
                dom: false,
                angular: false,
                react: false,
                vue: false,
            },
        },
        messages: {
            preferExplicitAssert: 'Wrap stand-alone `getBy*` query with `expect` function for better explicit assertion',
            preferExplicitAssertAssertion: '`getBy*` queries must be asserted with `{{assertion}}`',
        },
        schema: [
            {
                type: 'object',
                additionalProperties: false,
                properties: {
                    assertion: {
                        type: 'string',
                        enum: utils_1.PRESENCE_MATCHERS,
                    },
                },
            },
        ],
    },
    defaultOptions: [{}],
    create(context, [options], helpers) {
        const { assertion } = options;
        const getQueryCalls = [];
        return {
            'CallExpression Identifier'(node) {
                if (helpers.isGetQueryVariant(node)) {
                    getQueryCalls.push(node);
                }
            },
            'Program:exit'() {
                getQueryCalls.forEach((queryCall) => {
                    const node = node_utils_1.isMemberExpression(queryCall.parent)
                        ? queryCall.parent
                        : queryCall;
                    if (isAtTopLevel(node)) {
                        context.report({
                            node: queryCall,
                            messageId: 'preferExplicitAssert',
                        });
                    }
                    if (assertion) {
                        const expectCallNode = node_utils_1.findClosestCallNode(node, 'expect');
                        if (!expectCallNode)
                            return;
                        const expectStatement = expectCallNode.parent;
                        const property = expectStatement.property;
                        let matcher = property.name;
                        let isNegatedMatcher = false;
                        if (matcher === 'not' &&
                            node_utils_1.isMemberExpression(expectStatement.parent) &&
                            experimental_utils_1.ASTUtils.isIdentifier(expectStatement.parent.property)) {
                            isNegatedMatcher = true;
                            matcher = expectStatement.parent.property.name;
                        }
                        const shouldEnforceAssertion = (!isNegatedMatcher && utils_1.PRESENCE_MATCHERS.includes(matcher)) ||
                            (isNegatedMatcher && utils_1.ABSENCE_MATCHERS.includes(matcher));
                        if (shouldEnforceAssertion && matcher !== assertion) {
                            context.report({
                                node: property,
                                messageId: 'preferExplicitAssertAssertion',
                                data: {
                                    assertion,
                                },
                            });
                        }
                    }
                });
            },
        };
    },
});
