// react-mops/dist/index.js

/*!
 * @author Gregor Adams <greg@pixelass.com>
 * @name react-mops
 * @version 1.0.0
 * @see git+https://github.com/dekk-app/react-mops.git
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = _interopDefault(require('react'));
var cx = _interopDefault(require('classnames'));
var update = _interopDefault(require('immutability-helper'));
var uuid = require('uuid');

var styles = {"handles":"orZjL","forceHandle":"_2NQ23","drawOutline":"h0e5e","axis":"YKXop","guidesWrapper":"_3cZdS","guidesInner":"MAfLg","boundingBox":"_2rIYC","handleBase":"_29ZMP","full":"_22oV0","e":"_1f-Ay","n":"_3XR9A","ne":"_3qxAt","nw":"_2JuSc","s":"_3NfxN","se":"_3rzc7","sw":"_2ctb4","w":"_1O4_D","cursor":"K2UiS","handleMarker":"HJYgO","wrapper":"_2u2ae","content":"W61yA","move":"_2cd9s","weResize":"_1wf8l","nwseResize":"_2OuH3","nsResize":"_12o4C","neswResize":"_1I0YK","eRotate":"_3NtIy","nRotate":"_2cpdr","neRotate":"u0VtI","nwRotate":"_2X1JX","wRotate":"eLwNr","swRotate":"_2QIex","sRotate":"yll5l","seRotate":"Ipde_"};

const resizeClasses = [
    styles.colResize,
    styles.nwseResize,
    styles.rowResize,
    styles.neswResize
];
const rotationClasses = [
    styles.eRotate,
    styles.seRotate,
    styles.sRotate,
    styles.swRotate,
    styles.wRotate,
    styles.nwRotate,
    styles.nRotate,
    styles.neRotate
];

// tslint:disable-next-line:no-namespace
(function (Mops) {
    let HandleVariations;
    (function (HandleVariations) {
        HandleVariations[HandleVariations["e"] = 0] = "e";
        HandleVariations[HandleVariations["se"] = 1] = "se";
        HandleVariations[HandleVariations["s"] = 2] = "s";
        HandleVariations[HandleVariations["sw"] = 3] = "sw";
        HandleVariations[HandleVariations["w"] = 4] = "w";
        HandleVariations[HandleVariations["nw"] = 5] = "nw";
        HandleVariations[HandleVariations["n"] = 6] = "n";
        HandleVariations[HandleVariations["ne"] = 7] = "ne";
    })(HandleVariations = Mops.HandleVariations || (Mops.HandleVariations = {}));
})(exports.Mops || (exports.Mops = {}));

/*! *****************************************************************************
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the
License at http://www.apache.org/licenses/LICENSE-2.0

THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
MERCHANTABLITY OR NON-INFRINGEMENT.

See the Apache Version 2.0 License for specific language governing permissions
and limitations under the License.
***************************************************************************** */
function __rest(s, e) {
  var t = {};

  for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];

  if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
}

const handleVariations = [
    "e",
    "n",
    "ne",
    "nw",
    "s",
    "se",
    "sw",
    "w"
];
const handleDirs = {
    e: {
        x: 1,
        y: 0
    },
    n: {
        x: 0,
        y: -1
    },
    ne: {
        x: 1,
        y: -1
    },
    nw: {
        x: -1,
        y: -1
    },
    s: {
        x: 0,
        y: 1
    },
    se: {
        x: 1,
        y: 1
    },
    sw: {
        x: -1,
        y: 1
    },
    w: {
        x: -1,
        y: 0
    }
};
const Handle = React.forwardRef(({ fullSize, marker: Marker, onMouseDown, onTouchStart, variation }, ref) => (React.createElement("span", { ref: ref, className: cx("handleBase", variation, { [styles.full]: fullSize }), onMouseDown: onMouseDown, onTouchStart: onTouchStart }, Marker ? React.createElement(Marker, null) : null)));
const Handles = React.forwardRef((_a, ref) => {
    var { children, className, drawOutline } = _a, props = __rest(_a, ["children", "className", "drawOutline"]);
    return (React.createElement("div", Object.assign({}, props, { ref: ref, className: cx(className, styles.handles, {
            [styles.drawOutline]: drawOutline
        }) }), children));
});
Handle.defaultProps = {
    marker: null
};
const Wrapper = React.forwardRef((_a, ref) => {
    var { children, className, isDown, as: asElement } = _a, props = __rest(_a, ["children", "className", "isDown", "as"]);
    const As = asElement;
    return (React.createElement(As, Object.assign({}, props, { ref: ref, className: cx(className, styles.wrapper) }), children));
});
const Axis = React.forwardRef((_a, ref) => {
    var { children, className } = _a, props = __rest(_a, ["children", "className"]);
    return (React.createElement("div", Object.assign({}, props, { ref: ref, className: cx(className, styles.axis) }), children));
});
const GuidesWrapper = React.forwardRef((_a, ref) => {
    var { children, className } = _a, props = __rest(_a, ["children", "className"]);
    return (React.createElement("div", Object.assign({}, props, { ref: ref, className: cx(className, styles.guidesWrapper) }), children));
});
const GuidesInner = React.forwardRef((_a, ref) => {
    var { children, className } = _a, props = __rest(_a, ["children", "className"]);
    return (React.createElement("div", Object.assign({}, props, { ref: ref, className: cx(className, styles.guidesInner) }), children));
});
const Content = React.forwardRef((_a, ref) => {
    var { children, className } = _a, props = __rest(_a, ["children", "className"]);
    return (React.createElement("div", Object.assign({}, props, { ref: ref, className: cx(className, styles.content, {
            [styles.move]: typeof props.onMouseDown === "function"
        }) }), children));
});
const BoundingBox = React.forwardRef((_a, ref) => {
    var { children, className, drawOutline } = _a, props = __rest(_a, ["children", "className", "drawOutline"]);
    return (React.createElement("div", Object.assign({}, props, { ref: ref, className: cx(className, styles.boundingBox, {
            [styles.drawOutline]: drawOutline
        }) }), children));
});

const { Provider, Consumer } = React.createContext({
    // tslint:disable-next-line:no-empty
    addGuides: () => { },
    guideRequests: [],
    guides: [],
    // tslint:disable-next-line:no-empty
    hideGuides: () => { },
    // tslint:disable-next-line:no-empty
    removeGuides: () => { },
    // tslint:disable-next-line:no-empty
    showGuides: () => { },
    // tslint:disable-next-line:no-empty
    updateGuide: () => { }
});
const GuidesProvider = ({ children, guideRequests, containerSize }) => {
    const [guides, setGuides] = React.useState([]);
    const addGuides = React.useCallback(guideModels => {
        setGuides(state => {
            const newGuides = guideModels.filter(({ uuid }) => !Boolean(state.find(guide => {
                return guide.uuid === uuid;
            })));
            return update(state, {
                $push: newGuides
            });
        });
    }, [setGuides]);
    const removeGuides = React.useCallback(uuids => {
        setGuides(state => {
            return uuids
                ? update(state, {
                    $splice: uuids
                        .map(uuid => {
                        const index = state.findIndex(guide => guide.uuid === uuid);
                        if (index >= 0) {
                            return [index, 1];
                        }
                        return false;
                    })
                        .filter(Boolean)
                        .sort(([a], [b]) => b - a)
                })
                : [];
        });
    }, [setGuides]);
    const showGuides = React.useCallback(uuids => {
        setGuides(state => update(state, (uuids || state.map(({ uuid }) => uuid)).reduce((previousValue, currentValue) => {
            const index = state.findIndex(({ uuid }) => uuid === currentValue);
            return index > -1
                ? Object.assign(Object.assign({}, previousValue), { [index]: { visible: { $set: true } } }) : previousValue;
        }, {})));
    }, [setGuides]);
    const hideGuides = React.useCallback(uuids => {
        setGuides(state => update(state, (uuids || state.map(({ uuid }) => uuid)).reduce((previousValue, currentValue) => {
            const index = state.findIndex(({ uuid }) => uuid === currentValue);
            return index > -1
                ? Object.assign(Object.assign({}, previousValue), { [index]: { visible: { $set: false } } }) : previousValue;
        }, {})));
    }, [setGuides]);
    const updateGuide = React.useCallback(partialItem => {
        setGuides(state => {
            const index = state.findIndex(({ uuid }) => uuid === partialItem.uuid);
            return index > -1
                ? update(state, {
                    [index]: {
                        $merge: partialItem
                    }
                })
                : state;
        });
    }, [setGuides]);
    React.useEffect(() => {
        if (guideRequests) {
            const guideModels = guideRequests.map(({ uuid, x, y }) => {
                if (x !== undefined) {
                    return {
                        uuid,
                        x1: x,
                        x2: x,
                        y1: 0,
                        y2: containerSize.height
                    };
                }
                if (y !== undefined) {
                    return {
                        uuid,
                        x1: 0,
                        x2: containerSize.width,
                        y1: y,
                        y2: y
                    };
                }
            });
            addGuides(guideModels);
            const uuids = guideModels.map(({ uuid }) => uuid);
            return () => {
                removeGuides(uuids);
            };
        }
    }, [guideRequests, containerSize]);
    return (React.createElement(Provider, { value: {
            addGuides,
            guideRequests,
            guides,
            hideGuides,
            removeGuides,
            showGuides,
            updateGuide
        } }, children));
};
const Guides = React.forwardRef((_a, ref) => {
    var { guideColor, showGuides, height, width } = _a, props = __rest(_a, ["guideColor", "showGuides", "height", "width"]);
    const sizeRef = React.useRef();
    return (React.createElement(GuidesWrapper, { ref: ref },
        React.createElement(Consumer, null, ({ guides }) => {
            return (React.createElement(GuidesInner, { ref: sizeRef },
                React.createElement("svg", { viewBox: `0 0 ${width} ${height}` }, guides
                    .filter(({ visible }) => visible || showGuides)
                    .map((_a) => {
                    var { uuid, visible } = _a, guide = __rest(_a, ["uuid", "visible"]);
                    return (React.createElement("line", Object.assign({}, guide, { key: uuid, stroke: guideColor, strokeWidth: 2 })));
                }))));
        })));
});
Guides.defaultProps = {
    guideColor: "hsl(50, 100%, 50%)"
};

/**
 * @param value
 * @param altKey
 */
const withAlt = (value, altKey) => (altKey ? value * 2 : value);
/**
 * Determine the aspect-ratio from width and height
 * @param {Mops.SizeModel} size
 * @param {number} size.height
 * @param {number} size.width
 */
const getAspectRatio = ({ height, width }) => width / height;
/**
 * @param {number} value
 * @param {Mops.SizeModel} size
 * @param {boolean} reverse
 */
const withAspectRatio = (value, size, reverse) => reverse ? value / getAspectRatio(size) : value * getAspectRatio(size);
/**
 * Ensure degrees always ranges from 0 to < 360
 * @param deg
 */
const to360 = (deg) => (deg + 360) % 360;
const coordinatesToDeg = (position, center) => {
    const x = position.x - center.x;
    const y = position.y - center.y;
    return to360((Math.atan2(y, x) * 180) / Math.PI);
};
const normalize = n => {
    const rounded = Math.round(n * 1000000000) / 1000000000;
    if (rounded === 0 || rounded === -0) {
        return 0;
    }
    return n;
};
const polarToCartesian = (deg, radius = 1) => ({
    x: cos(deg) * radius,
    y: sin(deg) * radius
});
const cartesianToPolar = ({ x, y }) => ({
    deg: to360(atan2(y, x)),
    radius: getHypotenuse(y, x)
});
/**
 * Convert degrees to radians
 * @param {number} deg
 */
const degToRad = (deg) => (deg * Math.PI) / 180;
/**
 * Convert radians to degrees
 * @param {number} rad
 */
const radToDeg = (rad) => (rad * 180) / Math.PI;
/**
 * Determine the hypotenuse of a right triangle given the adjacent and opposite.
 * @param {number} opposite
 * @param {number} adjacent
 */
const getHypotenuse = (opposite, adjacent) => Math.sqrt(opposite * opposite + adjacent * adjacent);
/**
 * Math.cos using degrees instead of radians
 * @param {number} deg
 */
const cos = (deg) => Math.cos(degToRad(deg));
/**
 * Math.sin using degrees instead of radians
 * @param {number} deg
 */
const sin = (deg) => Math.sin(degToRad(deg));
/**
 * Math.atan2 returning degrees instead of radians
 * @param {number} opposite
 * @param {number} adjacent
 */
const atan2 = (opposite, adjacent) => radToDeg(Math.atan2(opposite, adjacent));
/**
 * Determine the transform values of a point with a rotation
 * @param {number} dX
 * @param {number} dY
 * @param {number} deg
 */
const withRotation = (dX, dY, deg) => {
    const hypotenuse = getHypotenuse(dX, dY);
    const beta = atan2(dY, dX);
    return polarToCartesian(beta - deg, hypotenuse);
};
/**
 * Get a rotation model from a mouseEvent and node, based on an existing rotation.
 * @param {HTMLElement} node
 * @param { React.MouseEvent} event
 * @param {{rotation:Mops.RotationModel,angle:Mops.RotationModel}} orientation
 */
const getRotation = (node, event, orientation) => {
    const { left, top, width, height } = node.getBoundingClientRect();
    const pointer = { x: event.clientX - left, y: event.clientY - top };
    const center = { x: width / 2, y: height / 2 };
    const deg = coordinatesToDeg(pointer, center);
    return {
        deg,
        rotation: (state) => ({
            x: state.x,
            y: state.y,
            z: to360(orientation.rotation.z + (deg - orientation.angle.z))
        })
    };
};
const getBounds = ({ height, width, angle }) => {
    const rad = degToRad(angle);
    const deg = radToDeg((rad > Math.PI * 0.5 && rad < Math.PI) || (rad > Math.PI * 1.5 && rad < Math.PI * 2)
        ? Math.PI - rad
        : rad);
    const s = sin(deg);
    const c = cos(deg);
    return {
        height: s * width + c * height,
        width: s * height + c * width
    };
};
const getInner = ({ height, width, angle }) => {
    const s = sin(angle);
    const c = cos(angle);
    const p = (1 / (c * c - s * s));
    return {
        height: p * (height * c - width * s),
        width: p * (width * c - height * s)
    };
};
const getBoundingBox = (m) => {
    const { height, width } = getBounds(m);
    return {
        height: Math.abs(height),
        width: Math.abs(width)
    };
};
const getInnerBox = (m) => {
    const { height, width } = getInner(m);
    return {
        height: Math.abs(height),
        width: Math.abs(width)
    };
};
const inRange = (value, min, max) => value >= min && value <= max;
const fallback = (...n) => n[0];
const chooseFn = (a, b = 0) => a > b ? Math.min : b > a ? Math.max : fallback;
const steps = (value, step) => Math.round(value / step) * step;

const toBounds = ({ top, right, bottom, left }) => ({ position, size, rotation }, {}, model = { position, size }) => {
    const boundingBox = getBoundingBox(Object.assign(Object.assign({}, size), { angle: rotation.z }));
    const boundaries = {
        bottom: bottom - boundingBox.height / 2,
        left: left + boundingBox.width / 2,
        right: right - boundingBox.width / 2,
        top: top + boundingBox.height / 2
    };
    return Object.assign(Object.assign({}, model), { position: {
            x: Math.max(boundaries.left, Math.min(boundaries.right, model.position.x)),
            y: Math.max(boundaries.top, Math.min(boundaries.bottom, model.position.y))
        } });
};
const toGrid = ({ x = 1, y = 1 }) => ({ position, size, rotation }, {}, model = { position, size }) => {
    const boundingBox = getBoundingBox(Object.assign(Object.assign({}, size), { angle: rotation.z }));
    const half = {
        x: boundingBox.width / 2,
        y: boundingBox.height / 2
    };
    const snap = {
        b: Math.round((model.position.y + half.y) / y) * y - half.y,
        l: Math.round((model.position.x - half.x) / x) * x + half.x,
        r: Math.round((model.position.x + half.x) / x) * x - half.x,
        t: Math.round((model.position.y - half.y) / y) * y + half.y
    };
    const diff = {
        b: Math.abs(model.position.y - snap.b),
        l: Math.abs(model.position.x - snap.l),
        r: Math.abs(model.position.x - snap.r),
        t: Math.abs(model.position.y - snap.t)
    };
    return Object.assign(Object.assign({}, model), { position: {
            x: diff.l < diff.r ? snap.l : snap.r,
            y: diff.t < diff.b ? snap.t : snap.b
        } });
};
const GUIDE_THRESHOLD = 10;
const toGuides = ({ threshold: { x: thresholdX = GUIDE_THRESHOLD, y: thresholdY = GUIDE_THRESHOLD } = {
    x: GUIDE_THRESHOLD,
    y: GUIDE_THRESHOLD
} }) => ({ position, size, rotation }, { guideRequests, guides, showGuides, hideGuides }, model = { position, size }) => {
    const tX = Math.max(GUIDE_THRESHOLD, thresholdX);
    const tY = Math.max(GUIDE_THRESHOLD, thresholdY);
    const withGuides = guideRequests.map(({ uuid, x, y }) => {
        const xMin = x - tX;
        const xMax = x + tX;
        const yMin = y - tY;
        const yMax = y + tY;
        const snapX = inRange(position.x, xMin, xMax);
        const snapY = inRange(position.y, yMin, yMax);
        const snap = { uuid, x: snapX && x, y: snapY && y };
        return snap;
    });
    const withSnap = withGuides.reduce((previousValue, { uuid, x, y }) => {
        const snap = previousValue;
        if (typeof x === "number") {
            snap.x = x;
            showGuides([uuid]);
        }
        else if (typeof y === "number") {
            snap.y = y;
            showGuides([uuid]);
        }
        else {
            hideGuides([uuid]);
        }
        return snap;
    }, {});
    return Object.assign(Object.assign({}, model), { position: Object.assign(Object.assign({}, model.position), withSnap) });
};
const SIBLING_X = uuid.v4();
const SIBLING_Y = uuid.v4();
const toSiblings = (siblings, { x: thresholdX = GUIDE_THRESHOLD, y: thresholdY = GUIDE_THRESHOLD } = {
    x: GUIDE_THRESHOLD,
    y: GUIDE_THRESHOLD
}) => ({ position, size, rotation }, { addGuides, removeGuides, updateGuide, guides, guideRequests, hideGuides }, model = { position, size }) => {
    const boundingBox = getBoundingBox(Object.assign(Object.assign({}, model.size), { angle: rotation.z }));
    const withBoundingBox = siblings.map(sibling => (Object.assign(Object.assign({}, sibling), { boundingBox: getBoundingBox({
            angle: sibling.rotation.z,
            height: sibling.size.height,
            width: sibling.size.width
        }) })));
    const initialValue = { x: {}, y: {} };
    const withSnap = withBoundingBox
        .map((_a) => {
        var { uuid } = _a, item = __rest(_a, ["uuid"]);
        return ({
            uuid,
            x: inRange(model.position.x, item.position.x - thresholdX, item.position.x + thresholdX)
                ? item.position.x
                : undefined,
            y: inRange(model.position.y, item.position.y - thresholdY, item.position.y + thresholdY)
                ? item.position.y
                : undefined
        });
    })
        .reduce((previousValue, { uuid, x, y }) => {
        const hadX = typeof previousValue.x.value === "number";
        const hadY = typeof previousValue.y.value === "number";
        const hasX = typeof x === "number";
        const hasY = typeof y === "number";
        const smallerX = hasX && hadX
            ? Math.abs(withBoundingBox.find(item => item.uuid === previousValue.x.uuid)
                .position.y - model.position.y) >
                Math.abs(withBoundingBox.find(item => item.uuid === uuid).position.y -
                    model.position.y)
            : true;
        const smallerY = hasY && hadY
            ? Math.abs(withBoundingBox.find(item => item.uuid === previousValue.y.uuid)
                .position.x - model.position.x) >
                Math.abs(withBoundingBox.find(item => item.uuid === uuid).position.x -
                    model.position.x)
            : true;
        return {
            x: {
                uuid: hasX && smallerX ? uuid : previousValue.x.uuid,
                value: hasX && smallerX ? x : previousValue.x.value
            },
            y: {
                uuid: hasY && smallerY ? uuid : previousValue.y.uuid,
                value: hasY && smallerY ? y : previousValue.y.value
            }
        };
    }, initialValue);
    const hasSnap = {
        x: typeof withSnap.x.value === "number",
        y: typeof withSnap.y.value === "number"
    };
    const snaplings = {
        x: hasSnap.x ? withBoundingBox.find(({ uuid }) => uuid === withSnap.x.uuid) : undefined,
        y: hasSnap.y ? withBoundingBox.find(({ uuid }) => uuid === withSnap.y.uuid) : undefined
    };
    if (hasSnap.x) {
        const dir = snaplings.x.position.y > model.position.y ? -1 : 1;
        const [y1, y2] = [
            snaplings.x.position.y - (snaplings.x.boundingBox.height / 2) * dir,
            (hasSnap.y ? snaplings.y.position.y : model.position.y) + (boundingBox.height / 2) * dir
        ];
        const guide = {
            uuid: SIBLING_X,
            visible: true,
            x1: withSnap.x.value,
            x2: withSnap.x.value,
            y1,
            y2
        };
        if (guides.find(({ uuid }) => uuid === SIBLING_X)) {
            updateGuide(guide);
        }
        else {
            addGuides([guide]);
        }
    }
    else {
        removeGuides([SIBLING_X]);
    }
    if (hasSnap.y) {
        const dir = snaplings.y.position.x > model.position.x ? -1 : 1;
        const [x1, x2] = [
            snaplings.y.position.x - (snaplings.y.boundingBox.width / 2) * dir,
            (hasSnap.x ? snaplings.x.position.x : model.position.x) + (boundingBox.width / 2) * dir
        ];
        const guide = {
            uuid: SIBLING_Y,
            visible: true,
            x1,
            x2,
            y1: withSnap.y.value,
            y2: withSnap.y.value
        };
        if (guides.find(({ uuid }) => uuid === SIBLING_Y)) {
            updateGuide(guide);
        }
        else {
            addGuides([guide]);
        }
    }
    else {
        removeGuides([SIBLING_Y]);
    }
    return Object.assign(Object.assign({}, model), { position: {
            x: hasSnap.x ? withSnap.x.value : model.position.x,
            y: hasSnap.y ? withSnap.y.value : model.position.y
        } });
};

const useAspectRatio = (aspectRatio, { setSize, setInitialSize, setPosition, setInitialPosition }) => {
    React.useEffect(() => {
        setSize(({ height }) => ({
            height,
            width: height * aspectRatio
        }));
    });
};

const useBoundingBox = (size, angle) => React.useMemo(() => getBoundingBox(Object.assign(Object.assign({}, size), { angle })), [size, angle]);

const useEventListener = (type, listener, context, dependencies) => {
    React.useEffect(() => {
        if (context !== null) {
            context.addEventListener(type, listener, { passive: true, capture: false });
        }
        return () => {
            if (context !== null) {
                context.removeEventListener(type, listener);
            }
        };
    }, [type, listener, context, ...dependencies]);
};
const useEventListeners = (types, listener, context, dependencies) => {
    React.useEffect(() => {
        if (context !== null) {
            types.map(type => context.addEventListener(type, listener, { passive: true, capture: false }));
        }
        return () => {
            if (context !== null) {
                types.map(type => context.removeEventListener(type, listener));
            }
        };
    }, [types, listener, context, ...dependencies]);
};

const useWindow = () => ("window" in global ? window : null);
const useDocument = () => ("window" in global && window.document ? document : null);

var KEYS;
(function (KEYS) {
    KEYS["Alt"] = "Alt";
    KEYS["Shift"] = "Shift";
    KEYS["Meta"] = "Meta";
    KEYS["Control"] = "Control";
})(KEYS || (KEYS = {}));
const useKey = key => {
    const [isActive, setActive] = React.useState(false);
    const win = useWindow();
    const doc = useDocument();
    useEventListener("keydown", (event) => {
        if (event.key === key) {
            setActive(true);
        }
    }, doc, [setActive, key]);
    // Deactivate on keyup
    useEventListener("keyup", (event) => {
        if (event.key === key) {
            setActive(false);
        }
    }, doc, [setActive, key]);
    // Deactivate on window blur and focus
    useEventListeners(["focus", "blur"], () => {
        setActive(false);
    }, win, [setActive]);
    return isActive;
};
const useMeta = () => useKey(KEYS.Meta);
const useControl = () => useKey(KEYS.Control);
const useShift = () => useKey(KEYS.Shift);
const useAlt = () => useKey(KEYS.Alt);

const usePointer = ({ onMouseMove, onMouseUp, onTouchEnd, onTouchMove }) => {
    const [isDown, setDown] = React.useState(false);
    if ("window" in global) {
        useEventListener("mousemove", (event) => {
            if (isDown) {
                onMouseMove(event);
            }
        }, document, [isDown, onMouseMove]);
        useEventListener("touchmove", (event) => {
            if (isDown) {
                onTouchMove(event);
            }
        }, document, [isDown, onTouchMove]);
        // Deactivate on mouseup
        useEventListener("mouseup", (event) => {
            if (isDown) {
                setDown(false);
                onMouseUp(event);
            }
        }, document, [isDown, setDown, onMouseUp]);
        // Deactivate on touchend
        useEventListener("touchend", (event) => {
            if (isDown) {
                setDown(false);
                onTouchEnd(event);
            }
        }, document, [isDown, setDown, onTouchEnd]);
        // Deactivate on window blur and focus
        useEventListeners(["focus", "blur"], () => {
            setDown(false);
        }, window, [setDown]);
        return setDown;
    }
    return (() => undefined);
};

const useCartesianToPolar = (callback, node, deps) => React.useCallback(({ x, y }, event) => {
    if (!node) {
        return;
    }
    if (x === undefined || y === undefined) {
        return callback();
    }
    const { top, left, height, width } = node.getBoundingClientRect();
    const { deg } = cartesianToPolar({
        x: x - left - width / 2,
        y: y - top - height / 2
    });
    callback(deg);
}, [node, callback, ...deps]);
const useRotation = (node, { onRotateStart, onRotate, onRotateEnd }, initialState = 0) => {
    const [initialRotation, setInitialRotation] = React.useState(initialState);
    const [rotation, setRotation] = React.useState(initialRotation);
    const handleMove = useCartesianToPolar(newRotation => {
        const deg = to360(newRotation - initialRotation);
        setRotation(deg);
        if (onRotate) {
            onRotate(deg);
        }
    }, node, [onRotate]);
    const handleUp = useCartesianToPolar((newRotation) => {
        const deg = newRotation !== undefined ? to360(newRotation - initialRotation) : rotation;
        setRotation(deg);
        if (onRotateEnd) {
            onRotateEnd(deg);
        }
    }, node, [onRotateEnd]);
    const onMouseMove = React.useCallback((event) => handleMove({ x: event.clientX, y: event.clientY }, event), [handleMove]);
    const onTouchMove = React.useCallback((event) => handleMove({ x: event.touches[0].clientX, y: event.touches[0].clientY }, event), [handleMove]);
    const handleDown = useCartesianToPolar(deg => {
        setInitialRotation(deg);
        if (onRotateStart) {
            onRotateStart(deg);
        }
    }, node, [onRotateStart]);
    const onMouseDown = React.useCallback((event) => {
        setDown(true);
        handleDown({ x: event.clientX, y: event.clientY }, event);
    }, [handleDown]);
    const onTouchStart = React.useCallback((event) => {
        setDown(true);
        handleDown({ x: event.touches[0].clientX, y: event.touches[0].clientY }, event);
    }, [handleDown]);
    const onMouseUp = React.useCallback((event) => {
        handleUp({ x: event.clientX, y: event.clientY }, event);
    }, [handleUp]);
    const onTouchEnd = React.useCallback((event) => {
        handleUp(rotation, event);
    }, [handleUp]);
    const setDown = usePointer({
        onMouseMove,
        onMouseUp,
        onTouchEnd,
        onTouchMove
    });
    return [rotation, { onMouseDown, onTouchStart }];
};

const useRotate = (initialState = 0, { steps: steps$1, step = 15, onRotateEnd, onRotate, onRotateStart }) => {
    const ref = React.useRef();
    const [initialAngle, setInitialAngle] = React.useState(initialState);
    const [angle, setAngle] = React.useState(initialAngle);
    const handleRotate = React.useCallback(d => {
        const newAngle = to360(steps(initialAngle + d, step && steps$1 ? step : 1));
        setAngle(newAngle);
        if (onRotate) {
            onRotate({ rotation: { x: 0, y: 0, z: newAngle } });
        }
    }, [step, steps$1, setAngle, initialAngle]);
    const handleRotateEnd = React.useCallback(d => {
        const newAngle = to360(steps(initialAngle + d, step && steps$1 ? step : 1));
        setAngle(newAngle);
        setInitialAngle(newAngle);
        if (onRotateEnd) {
            onRotateEnd({ rotation: { x: 0, y: 0, z: newAngle } });
        }
    }, [step, steps$1, setAngle, initialAngle, setInitialAngle, onRotateEnd]);
    const handleRotateStart = React.useCallback(() => {
        if (onRotateStart) {
            onRotateStart({ rotation: { x: 0, y: 0, z: initialAngle } });
        }
    }, [onRotateStart]);
    const [, methods] = useRotation(ref.current, {
        onRotate: handleRotate,
        onRotateEnd: handleRotateEnd,
        onRotateStart: handleRotateStart
    });
    return [angle, Object.assign(Object.assign({}, methods), { ref })];
};

const useOffset = ({ onDragStart, onDrag, onDragEnd }, initialState = { x: 0, y: 0 }) => {
    const [initialOffset, setInitialOffset] = React.useState(initialState);
    const [offset, setOffset] = React.useState(initialOffset);
    const handleMove = React.useCallback(pointer => {
        const coords = {
            x: pointer.x - initialOffset.x,
            y: pointer.y - initialOffset.y
        };
        setOffset(coords);
        onDrag(coords);
    }, [setOffset, initialOffset, onDrag]);
    const onMouseMove = React.useCallback((event) => {
        handleMove({ x: event.clientX, y: event.clientY });
    }, [handleMove]);
    const onTouchMove = React.useCallback((event) => {
        handleMove({ x: event.touches[0].clientX, y: event.touches[0].clientY });
    }, [handleMove]);
    const handleUp = React.useCallback((pointer) => {
        const coords = pointer
            ? {
                x: pointer.x - initialOffset.x,
                y: pointer.y - initialOffset.y
            }
            : offset;
        setOffset(coords);
        onDragEnd(coords);
    }, [setOffset, initialOffset, onDragEnd]);
    const onMouseUp = React.useCallback((event) => {
        handleUp({ x: event.clientX, y: event.clientY });
    }, [handleUp]);
    const onTouchEnd = React.useCallback((event) => {
        handleUp();
    }, [handleUp]);
    const handleDown = React.useCallback(pointer => {
        setDown(true);
        setInitialOffset(pointer);
        onDragStart({ x: 0, y: 0 });
    }, [setInitialOffset, onDragStart]);
    const onMouseDown = React.useCallback((event) => {
        handleDown({ x: event.clientX, y: event.clientY });
    }, [handleDown]);
    const onTouchStart = React.useCallback((event) => {
        handleDown({ x: event.touches[0].clientX, y: event.touches[0].clientY });
    }, [handleDown]);
    const setDown = usePointer({
        onMouseMove,
        onMouseUp,
        onTouchEnd,
        onTouchMove
    });
    return [offset, { onMouseDown, onTouchStart }];
};

const usePosition = (initialState = { x: 0, y: 0 }, { onDragEnd, onDragStart, onDrag }) => {
    const [initialPosition, setInitialPosition] = React.useState(initialState);
    const [position, setPosition] = React.useState(initialPosition);
    const handleDrag = React.useCallback(o => {
        const newPosition = {
            x: initialPosition.x + o.x,
            y: initialPosition.y + o.y
        };
        setPosition(newPosition);
        if (onDrag) {
            onDrag({ position: newPosition });
        }
    }, [setPosition, initialPosition, onDrag]);
    const handleDragEnd = React.useCallback(o => {
        const newPosition = {
            x: initialPosition.x + o.x,
            y: initialPosition.y + o.y
        };
        setPosition(newPosition);
        setInitialPosition(newPosition);
        if (onDragEnd) {
            onDragEnd({ position: newPosition });
        }
    }, [setPosition, setInitialPosition, initialPosition, onDragEnd]);
    const handleDragStart = React.useCallback(() => {
        if (onDragStart) {
            onDragStart({ position: initialPosition });
        }
    }, [onDragStart]);
    const [, methods] = useOffset({
        onDrag: handleDrag,
        onDragEnd: handleDragEnd,
        onDragStart: handleDragStart
    });
    return [position, Object.assign(Object.assign({}, methods), { initialPosition, setInitialPosition, setPosition })];
};

const useSize = (initialState, { centered, deg = 0, dir = { x: 1, y: 1 }, onResize, onResizeEnd, onResizeStart, initialPosition, setInitialPosition, setPosition }) => {
    const [initialSize, setInitialSize] = React.useState(initialState);
    const [size, setSize] = React.useState(initialSize);
    const setValues = React.useCallback(({ x, y }, end = false) => {
        const { x: rX, y: rY } = withRotation(x, y, deg);
        const sizeOffset = {
            height: initialSize.height + rY * dir.y * (centered ? 2 : 1),
            width: initialSize.width + rX * dir.x * (centered ? 2 : 1)
        };
        const newSize = {
            height: Math.abs(sizeOffset.height),
            width: Math.abs(sizeOffset.width)
        };
        const { x: dX, y: dY } = withRotation(centered || dir.x === 0 ? 0 : rX, centered || dir.y === 0 ? 0 : rY, -deg);
        const newPosition = {
            x: initialPosition.x + dX / 2,
            y: initialPosition.y + dY / 2
        };
        if (setPosition) {
            setPosition(newPosition);
        }
        setSize(newSize);
        if (end) {
            setInitialSize(newSize);
            if (setInitialPosition) {
                setInitialPosition(newPosition);
            }
            if (onResizeEnd) {
                onResizeEnd({
                    size: newSize
                });
            }
        }
        else {
            if (onResize) {
                onResize({
                    size: newSize
                });
            }
        }
    }, [
        centered,
        initialSize,
        setSize,
        setInitialSize,
        dir,
        deg,
        setPosition,
        onResizeEnd,
        onResize,
        initialPosition,
        setInitialPosition
    ]);
    const handleDrag = React.useCallback((offset) => {
        setValues(offset);
    }, [setValues]);
    const handleDragEnd = React.useCallback((offset) => {
        setValues(offset, true);
    }, [setValues]);
    const handleDragStart = React.useCallback(() => {
        if (onResizeStart) {
            onResizeStart({
                size: initialSize
            });
        }
    }, [onResizeStart]);
    const [, methods] = useOffset({
        onDrag: handleDrag,
        onDragEnd: handleDragEnd,
        onDragStart: handleDragStart
    });
    return [size, Object.assign(Object.assign({}, methods), { setSize, setInitialSize })];
};

const useSnap = (shouldSnap, { position, rotation, size }, guidesContext) => React.useMemo(() => shouldSnap.reduce((model, fn) => {
    const update = fn({
        position,
        rotation,
        size
    }, guidesContext, model);
    return {
        position: Object.assign(Object.assign({}, model.position), update.position),
        rotation,
        size: Object.assign(Object.assign({}, model.size), update.size)
    };
}, { position, rotation, size }), [position, rotation, shouldSnap, size, guidesContext]);

const useViewport = ({ fallbackSize, onResize }) => {
    const win = useWindow();
    const [size, setSize] = React.useState(fallbackSize);
    React.useEffect(() => {
        const newSize = {
            height: win ? win.innerHeight : fallbackSize.height,
            width: win ? win.innerWidth : fallbackSize.width
        };
        setSize(newSize);
    }, [fallbackSize]);
    useEventListener("resize", () => {
        const newSize = {
            height: win ? win.innerHeight : fallbackSize.height,
            width: win ? win.innerWidth : fallbackSize.width
        };
        setSize(newSize);
        if (onResize) {
            onResize(newSize);
        }
    }, win, [fallbackSize, setSize, onResize]);
    return size;
};

var OS;
(function (OS) {
    OS["OSX"] = "OSX";
    OS["WINDOWS"] = "WINDOWS";
    OS["LINUX"] = "LINUX";
    OS["UNIX"] = "UNIX";
    OS["NODE"] = "NODE";
    OS["Win"] = "Win";
    OS["Mac"] = "Mac";
    OS["Unix"] = "X11";
    OS["Linux"] = "Linux";
})(OS || (OS = {}));
const hasGlobalProp = prop => prop in global;
const getOS = () => {
    if (hasGlobalProp("navigator")) {
        if (navigator.appVersion.indexOf(OS.Win) !== -1) {
            return OS.WINDOWS;
        }
        if (navigator.appVersion.indexOf(OS.Mac) !== -1) {
            return OS.OSX;
        }
        if (navigator.appVersion.indexOf(OS.Unix) !== -1) {
            return OS.UNIX;
        }
        if (navigator.appVersion.indexOf(OS.Linux) !== -1) {
            return OS.LINUX;
        }
    }
    return OS.NODE;
};
const isOSX = () => getOS() === OS.OSX;

const Box = ({ as, children, drawBoundingBox, drawBox, fullHandles, size: initialSize, position: initialPosition, rotation: initialRotation, onRotateEnd, onDragEnd, onResizeEnd, onRotateStart, onDragStart, onResizeStart, onRotate, onDrag, onResize, isResizable, isDraggable, marker, isRotatable, shouldSnap = [], addGuides, guideRequests, guides, hideGuides, removeGuides, showGuides, updateGuide }) => {
    const [isActive, setActive] = React.useState(false);
    const [dir, setDir] = React.useState({ x: 0, y: 0 });
    const metaKey = useMeta();
    const controlKey = useControl();
    const altKey = useAlt();
    const shiftKey = useShift();
    const [rotate, rotateProps] = useRotate(initialRotation.z, {
        onRotate,
        onRotateEnd: b => {
            const update = onEnd();
            if (onRotateEnd) {
                onRotateEnd(Object.assign(Object.assign({}, b), update));
            }
        },
        onRotateStart: b => {
            if (onRotateStart) {
                onRotateStart(b);
            }
        },
        step: 15,
        steps: shiftKey
    });
    const [position, positionProps] = usePosition(initialPosition, {
        onDrag,
        onDragEnd: o => {
            setActive(false);
            const update = onEnd();
            if (onDragEnd) {
                onDragEnd(Object.assign(Object.assign({}, o), update));
            }
        },
        onDragStart: b => {
            setActive(true);
            if (onDragStart) {
                onDragStart(b);
            }
        }
    });
    const [size, sizeProps] = useSize(initialSize, {
        centered: altKey,
        deg: rotate,
        dir,
        initialPosition: positionProps.initialPosition,
        onResize,
        onResizeEnd: b => {
            const update = onEnd();
            if (onResizeEnd) {
                onResizeEnd(Object.assign(Object.assign({}, b), update));
            }
        },
        onResizeStart: b => {
            if (onResizeStart) {
                onResizeStart(b);
            }
        },
        setInitialPosition: positionProps.setInitialPosition,
        setPosition: positionProps.setPosition
    });
    const snap = useSnap(isActive ? shouldSnap : [], { size, position, rotation: Object.assign(Object.assign({}, initialRotation), { z: rotate }) }, { addGuides, guideRequests, guides, hideGuides, removeGuides, showGuides, updateGuide });
    const onEnd = React.useCallback(() => {
        const newPosition = snap.position;
        const newSize = snap.size;
        positionProps.setPosition(newPosition);
        positionProps.setInitialPosition(newPosition);
        sizeProps.setSize(newSize);
        sizeProps.setInitialSize(newSize);
        hideGuides();
        return {
            position: newPosition,
            size: newSize
        };
    }, [hideGuides, sizeProps, positionProps]);
    const boundingBox = useBoundingBox(snap.size, rotate);
    console.log(guides);
    return (React.createElement(Wrapper, { as: as, style: {
            transform: `translate3d(${snap.position.x}px,${snap.position.y}px,0) translate3d(-50%,-50%, 0)`
        } },
        React.createElement(BoundingBox, { drawOutline: drawBoundingBox, style: boundingBox, ref: rotateProps.ref },
            React.createElement(Axis, { style: {
                    transform: `translate3d(-50%,-50%, 0) rotate3d(0,0,1,${rotate}deg)`
                } },
                isResizable || isRotatable ? (React.createElement(Handles, { drawOutline: drawBox }, handleVariations.map(key => (React.createElement(Handle, { key: key, variation: key, marker: marker, fullSize: fullHandles, onMouseDown: event => {
                        setDir(handleDirs[key]);
                        if ((isOSX() ? metaKey : controlKey) && isRotatable) {
                            rotateProps.onMouseDown(event);
                        }
                        else if (isResizable) {
                            sizeProps.onMouseDown(event);
                        }
                    }, onTouchStart: event => {
                        setDir(handleDirs[key]);
                        if (isResizable) {
                            sizeProps.onTouchStart(event);
                        }
                    } }))))) : null,
                React.createElement(Content, { style: Object.assign({}, snap.size), onMouseDown: isDraggable ? positionProps.onMouseDown : undefined, onTouchStart: event => {
                        if (event.touches.length > 1) {
                            if (isRotatable) {
                                rotateProps.onTouchStart(event);
                            }
                        }
                        else if (isDraggable) {
                            positionProps.onTouchStart(event);
                        }
                    } }, children)))));
};
Box.defaultProps = {
    as: "div",
    marker: null
};
const GuidedBox = (_a) => {
    var { children } = _a, props = __rest(_a, ["children"]);
    return (React.createElement(Consumer, null, context => (React.createElement(Box, Object.assign({}, props, context), children))));
};

exports.GuidedBox = GuidedBox;
exports.Guides = Guides;
exports.GuidesConsumer = Consumer;
exports.GuidesProvider = GuidesProvider;
exports.atan2 = atan2;
exports.cartesianToPolar = cartesianToPolar;
exports.chooseFn = chooseFn;
exports.coordinatesToDeg = coordinatesToDeg;
exports.cos = cos;
exports.degToRad = degToRad;
exports.getBoundingBox = getBoundingBox;
exports.getBounds = getBounds;
exports.getHypotenuse = getHypotenuse;
exports.getInner = getInner;
exports.getInnerBox = getInnerBox;
exports.getRotation = getRotation;
exports.inRange = inRange;
exports.normalize = normalize;
exports.polarToCartesian = polarToCartesian;
exports.radToDeg = radToDeg;
exports.resizeClasses = resizeClasses;
exports.rotationClasses = rotationClasses;
exports.sin = sin;
exports.steps = steps;
exports.to360 = to360;
exports.toBounds = toBounds;
exports.toGrid = toGrid;
exports.toGuides = toGuides;
exports.toSiblings = toSiblings;
exports.useAlt = useAlt;
exports.useAspectRatio = useAspectRatio;
exports.useBoundingBox = useBoundingBox;
exports.useControl = useControl;
exports.useEventListener = useEventListener;
exports.useEventListeners = useEventListeners;
exports.useKey = useKey;
exports.useMeta = useMeta;
exports.usePosition = usePosition;
exports.useRotate = useRotate;
exports.useShift = useShift;
exports.useSize = useSize;
exports.useSnap = useSnap;
exports.useViewport = useViewport;
exports.withAlt = withAlt;
exports.withAspectRatio = withAspectRatio;
exports.withRotation = withRotation;
